contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,508 @@
+            execute_register_merkle_root(deps, env, info, merkle_root)
+        }
+        ExecuteMsg::Claim {
+            stage,
+            amount,
+            proof,
+        } => execute_claim(deps, env, info, stage, amount, proof),
+    }
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    owner: Option<String>,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // validate owner change
+    let new_owner = owner
+        .ok_or(ContractError::InvalidInput {})
+        .and_then(|o| deps.api.addr_validate(o.as_str()).map_err(|err| err.into()))?;
+
+    CONFIG.update(deps.storage, |mut exists| -> StdResult<_> {
+        exists.owner = new_owner;
+        Ok(exists)
+    })?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![attr("action", "update_config")],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_register_merkle_root(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    merkle_root: String,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // check merkle root length
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root.to_string(), &mut root_buf)?;
+
+    let latest_stage: u8 = STAGE.load(deps.storage)?;
+    let stage = latest_stage + 1;
+
+    MERKLE_ROOT.save(deps.storage, U8Key::from(stage), &merkle_root)?;
+    STAGE.save(deps.storage, &stage)?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![
+            attr("action", "register_merkle_root"),
+            attr("stage", stage),
+            attr("merkle_root", merkle_root),
+        ],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_claim(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    stage: u8,
+    amount: Uint128,
+    proof: Vec<String>,
+) -> Result<Response, ContractError> {
+    // verify not claimed
+    let claimed = CLAIM.may_load(deps.storage, (&info.sender, U8Key::from(stage)))?;
+    if claimed.is_some() {
+        return Err(ContractError::Claimed {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+    let merkle_root = MERKLE_ROOT.load(deps.storage, stage.into())?;
+
+    let user_input: String = info.sender.to_string() + &amount.to_string();
+    let mut hash: [u8; 32] = sha3::Keccak256::digest(user_input.as_bytes())
+        .as_slice()
+        .try_into()
+        .map_err(|_| ContractError::WrongLength {})?;
+
+    for p in proof {
+        let mut proof_buf: [u8; 32] = [0; 32];
+        hex::decode_to_slice(p, &mut proof_buf)?;
+
+        hash = match bytes_cmp(hash, proof_buf) {
```

Comment:
I would convert it to something like:
```
let mut hashes = [hash, proof_buf];
hashes.sort();
sha3::Keccak256::digest(&hashes.concat())
  .as_slice().
  .try_into()
  .map_err(|_| ContractError::WrongLength {})?;
```

Honestly - I spend way too much time on looking for difference in those branches.

Also this whole loop looks very much as fold to me, which I find way more readable than mutating variable:

```
let hash = proof.into_iter().try_fold(hash, |hash, p| {
  let mut proof_buf = [0; 32];
  hex::decode_to_slice(p, &mut proof_buf)?;
  let mut hashes = [hash, proof_buf];
  hashes.sort();
  sha3::Keccak256::digest(&hashes.concat())
    .as_slice()
    .try_into()
    .map_err(|_| ConcrtactError::WrongLength {})
})?;
```

Reply:
Good one

Reply:
This is thrown by linter:
```
error: used `sort` on primitive type `array`
   --> contracts/cw20-merkle-airdrop/src/contract.rs:154:9
    |
154 |         hashes.sort();
    |         ^^^^^^^^^^^^^ help: try: `hashes.sort_unstable()`
    |
    = note: `-D clippy::stable-sort-primitive` implied by `-D warnings`
    = note: an unstable sort would perform faster without any observable difference for this data type
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#stable_sort_primitive
```

Reply:
Clippy is right here - `sort_unstable` is faster, and for primitive types order by definition doesn't matter for equal elements (as equal is "the same" for them). Just change `sort` to `sort_unstable`.